File: annotationprocessor/src/main/java/com/quarkworks/android/realmtypesafequery/annotationprocessor/Maps.java
package com.quarkworks.android.realmtypesafequery.annotationprocessor;


import com.squareup.javapoet.ClassName;

import java.util.HashMap;

/**
 * Created by james on 9/12/2016.
 */

class Maps {
    public static final HashMap<String, ClassName> BaseMap;
    public static final HashMap<ClassName, ClassName> IndexMap;

    public static final ClassName realmbooleanfield    = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmBooleanField")  ;
    public static final ClassName realmbytefield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmByteField")     ;
    public static final ClassName realmshortfield      = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmShortField")    ;
    public static final ClassName realmintegerfield    = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmIntegerField")  ;
    public static final ClassName realmlongfield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmLongField")     ;
    public static final ClassName realmfloatfield      = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmFloatField")    ;
    public static final ClassName realmdoublefield     = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmDoubleField")   ;
    public static final ClassName realmstringfield     = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmStringField")   ;
    public static final ClassName realmbytearrayfield  = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmByteArrayField");
    public static final ClassName realmdatefield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields", "RealmDateField")     ;

    public static final ClassName realmindexedbooleanfield    = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedBooleanField")  ;
    public static final ClassName realmindexedbytefield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedByteField")     ;
    public static final ClassName realmindexedshortfield      = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedShortField")    ;
    public static final ClassName realmindexedintegerfield    = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedIntegerField")  ;
    public static final ClassName realmindexedlongfield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedLongField")     ;
    public static final ClassName realmindexedstringfield     = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedStringField")   ;
    public static final ClassName realmindexeddatefield       = ClassName.get("com.quarkworks.android.realmtypesafequery.fields.indexed", "RealmIndexedDateField")     ;

    public static final ClassName realmtoonerelationship       = ClassName.get("com.quarkworks.android.realmtypesafequery.relationships", "RealmToOneRelationship")    ;
    public static final ClassName realmtomanyrelationship      = ClassName.get("com.quarkworks.android.realmtypesafequery.relationships", "RealmToManyRelationship")   ;

    static {
        BaseMap = new HashMap<>();

        BaseMap.put("java.lang.Boolean",  realmbooleanfield   );
        BaseMap.put("java.lang.Byte",     realmbytefield      );
        BaseMap.put("java.lang.Short",    realmshortfield     );
        BaseMap.put("java.lang.Integer",  realmintegerfield   );
        BaseMap.put("java.lang.Long",     realmlongfield      );
        BaseMap.put("java.lang.Float",    realmfloatfield     );
        BaseMap.put("java.lang.Double",   realmdoublefield    );
        BaseMap.put("java.lang.String",   realmstringfield    );
        BaseMap.put("byte[]",             realmbytearrayfield );
        BaseMap.put("java.util.Date",     realmdatefield      );

        BaseMap.put("boolean",            realmbooleanfield   );
        BaseMap.put("byte",               realmbytefield      );
        BaseMap.put("short",              realmshortfield     );
        BaseMap.put("int",                realmintegerfield   );
        BaseMap.put("long",               realmlongfield      );
        BaseMap.put("float",              realmfloatfield     );
        BaseMap.put("double",             realmdoublefield    );


        IndexMap = new HashMap<>();

        IndexMap.put(  realmbooleanfield  , realmindexedbooleanfield );
        IndexMap.put(  realmbytefield     , realmindexedbytefield    );
        IndexMap.put(  realmshortfield    , realmindexedshortfield   );
        IndexMap.put(  realmintegerfield  , realmindexedintegerfield );
        IndexMap.put(  realmlongfield     , realmindexedlongfield    );
        IndexMap.put(  realmstringfield   , realmindexedstringfield  );
        IndexMap.put(  realmdatefield     , realmindexeddatefield    );
    }
}

File: annotationprocessor/src/main/java/com/quarkworks/android/realmtypesafequery/annotationprocessor/Maps.kt
package com.quarkworks.android.realmtypesafequery.annotationprocessor

import com.squareup.javapoet.ClassName

import java.util.HashMap

internal object Maps {
    val BASE_MAP = HashMap<String, ClassName>()
    val INDEX_MAP = HashMap<ClassName, ClassName>()
    
    private val PACKAGE = "com.quarkworks.android.realmtypesafequery.fields"
    
    private fun getClassName(simpleName: String) : ClassName = ClassName.get(PACKAGE, simpleName)!!

    private val REALM_BOOLEAN_FIELD = getClassName("RealmBooleanField")
    private val REALM_BYTE_FIELD = getClassName("RealmByteField")
    private val REALM_SHORT_FIELD = getClassName("RealmShortField")
    private val REALM_INTEGER_FIELD = getClassName("RealmIntegerField")
    private val REALM_LONG_FIELD = getClassName("RealmLongField")
    private val REALM_FLOAT_FIELD = getClassName("RealmFloatField")
    private val REALM_DOUBLE_FIELD = getClassName("RealmDoubleField")
    private val REALM_STRING_FIELD = getClassName("RealmStringField")
    private val REALM_BYTE_ARRAY_FIELD = getClassName("RealmByteArrayField")
    private val REALM_DATE_FIELD = getClassName("RealmDateField")

    private val REALM_INDEXED_BOOLEAN_FIELD = getClassName("RealmIndexedBooleanField")
    private val REALM_INDEXED_BYTE_FIELD = getClassName("RealmIndexedByteField")
    private val REALM_INDEXED_SHORT_FIELD = getClassName("RealmIndexedShortField")
    private val REALM_INDEXED_INTEGER_FIELD = getClassName("RealmIndexedIntegerField")
    private val REALM_INDEXED_LONG_FIELD = getClassName("RealmIndexedLongField")
    private val REALM_INDEXED_STRING_FIELD = getClassName("RealmIndexedStringField")
    private val REALM_INDEXED_DATE_FIELD = getClassName("RealmIndexedDateField")

    val REALM_TO_ONE_RELATIONSHIP = getClassName("RealmToOneRelationship")
    val REALM_TO_MANY_RELATIONSHIP = getClassName("RealmToManyRelationship")

    init {
        BASE_MAP.put("java.lang.Boolean", REALM_BOOLEAN_FIELD)
        BASE_MAP.put("java.lang.Byte", REALM_BYTE_FIELD)
        BASE_MAP.put("java.lang.Short", REALM_SHORT_FIELD)
        BASE_MAP.put("java.lang.Integer", REALM_INTEGER_FIELD)
        BASE_MAP.put("java.lang.Long", REALM_LONG_FIELD)
        BASE_MAP.put("java.lang.Float", REALM_FLOAT_FIELD)
        BASE_MAP.put("java.lang.Double", REALM_DOUBLE_FIELD)
        BASE_MAP.put("java.lang.String", REALM_STRING_FIELD)
        BASE_MAP.put("java.util.Date", REALM_DATE_FIELD)

        BASE_MAP.put("boolean", REALM_BOOLEAN_FIELD)
        BASE_MAP.put("byte", REALM_BYTE_FIELD)
        BASE_MAP.put("byte[]", REALM_BYTE_ARRAY_FIELD)
        BASE_MAP.put("short", REALM_SHORT_FIELD)
        BASE_MAP.put("int", REALM_INTEGER_FIELD)
        BASE_MAP.put("long", REALM_LONG_FIELD)
        BASE_MAP.put("float", REALM_FLOAT_FIELD)
        BASE_MAP.put("double", REALM_DOUBLE_FIELD)

        INDEX_MAP.put(REALM_BOOLEAN_FIELD, REALM_INDEXED_BOOLEAN_FIELD)
        INDEX_MAP.put(REALM_BYTE_FIELD, REALM_INDEXED_BYTE_FIELD)
        INDEX_MAP.put(REALM_SHORT_FIELD, REALM_INDEXED_SHORT_FIELD)
        INDEX_MAP.put(REALM_INTEGER_FIELD, REALM_INDEXED_INTEGER_FIELD)
        INDEX_MAP.put(REALM_LONG_FIELD, REALM_INDEXED_LONG_FIELD)
        INDEX_MAP.put(REALM_STRING_FIELD, REALM_INDEXED_STRING_FIELD)
        INDEX_MAP.put(REALM_DATE_FIELD, REALM_INDEXED_DATE_FIELD)
    }
}
